### PR TITLE
[MRG] compute yscale from cell points

### LINKED ISSUE #326: compute cos theta from actual section points
Currently `yscale` (cos theta) is hardcoded when computing the dipole. See

https://github.com/jonescompneurolab/hnn-core/blob/e0377477363185c1283514889d1c4f34afb61af2/hnn_core/cell.py#L213

and

https://github.com/jonescompneurolab/hnn-core/blob/e0377477363185c1283514889d1c4f34afb61af2/hnn_core/params_default.py#L415-L424

what needs to happen is that `sec_pts` should be used for the `apical_trunk` segment and the current section and then cos theta should be computed as something like:

```py
import numpy as np
from numpy.linalg import norm

yscales = dict()
for sec_name in p_secs:
    a = p_secs['apical_trunk']['sec_pts'][1] - p_secs['apical_trunk']['sec_pts'][0]
    b = p_secs[sec_name]['sec_pts'][1] - p_secs[sec_name]['sec_pts'][0]
    yscales[sec_name] = np.dot(a, b) / (norm(a) * norm(b))
```

maybe the section name corresponding to the trunk should be an argument to `insert_dipole`

rainy sunday afternoon hack for someone bored ...

### PR DESCRIPTION
**Pull request by jasmainak**
closes #326 

I tried not to touch too much so others can continue with their PRs

### CODE REVIEW

File: hnn_core/cell.py
Comment (jasmainak) [inline hnn_core/cell.py:309]: it's becoming clear that `p_secs` should be an attribute of `Cell` but I'm holding off on that in this PR.

Comment (rythorpe) [reply hnn_core/cell.py:309]: I don't see why `p_secs` needs to be an attribute of `Cell` since it isn't used anywhere outside `Cell.build()` and here. Since `Cell.insert_dipole()` is called after `build()`, why not just access the points via the NEURON Section objects?

File: hnn_core/cell.py
Comment (rythorpe) [inline hnn_core/cell.py:329]: Re: previous comment, `_get_yscale()` could return the y-scale for a single section called within the for loop below at L334.

File: hnn_core/params_default.py
Comment (rythorpe) [inline hnn_core/params_default.py:423]: If I recall correctly, the true morphology of these basal dendrites doesn't orient them at 45 degrees to the apical dendrite. How is the calculated dipole in this PR still matching the ground-truth dipole in our tests?

Comment (jasmainak) [reply hnn_core/params_default.py:423]: It does orient at 45 degrees actually. Take a look at the points above:

```py
'apical_trunk': [[-50, 778, 0], [-50, 813, 0]],
'basal_3': [[-50, 715, 0], [56, 609, 0]],
```

apical -> 35j
basal_3 -> 106i + 106j

First vector is parallel to y axis and second one has equal components in x and y axis. Thus they are at 45 degrees.

File: hnn_core/cell.py
Comment (jasmainak) [reply hnn_core/cell.py:309]: I'm imagining that `plot_cell_morphology` will move from `Network` to `Cell` and then `self.p_secs` will be used for the plot. I'm a little wary of relying on NEURON objects because of : 1) numerical issues we saw in the past 2) unpicklability. Since `p_secs` is picklable we can make it accessible to the user to inspect/modify etc.

File: hnn_core/params_default.py
Comment (rythorpe) [reply hnn_core/params_default.py:423]: Oh okay, so were the axes in `plot_cell_morphology` not equivalent when we were reviewing this the other day?

Comment (jasmainak) [reply hnn_core/params_default.py:423]: optical illusion I would say :-) Because the apical dendrite is longer than the basal, it looked as if the angle was smaller than it really was

File: hnn_core/cell.py
Comment (rythorpe) [reply hnn_core/cell.py:309]: Fair enough!

File: doc/whats_new.rst
Comment (cjayb) [inline doc/whats_new.rst:30]: I realised right after merging 322 that we missed this...

File: hnn_core/cell.py
Comment (cjayb) [reply hnn_core/cell.py:329]: I'm confused about this too! I think variable & function naming should be better. "yscale" is actually a projection of each (linear) section onto the line defined by the orientation of the apical dendrite. How about:

```python
apical_projs = _get_sect_projs_onto_apical_orientation(p_secs, orientation='apical_trunk')
```

File: hnn_core/cell.py
Comment (cjayb) [inline hnn_core/cell.py:15]: See comment below, proposing to call this something to do with the projection that it is. It should be explicit that `dipole_sec` _defines_ the orientation. I'd prefer we don't use `yscale` anywhere, however...

File: hnn_core/cell.py
Comment (cjayb) [inline hnn_core/cell.py:17]: Here we have the hard assumption that our cells live in the xy-plane. Not much to do about it, just a note.

File: hnn_core/cell.py
Comment (cjayb) [reply hnn_core/cell.py:310]: I agree that to enable `Cell.plot` we should keep HOC-objects out of ours for as long as possible.

File: hnn_core/cell.py
Comment (cjayb) [inline hnn_core/cell.py:347]: This (`y_scale` and `yscale`) is just unfair! We have to change these variable names while we're at it :)

```python
# Dipoles are oriented along the section. We record the projection of the
# current dipole moment onto the direction of the apical dendrite
dip_proj = (apical_projs[sect_name] * sect.L) * pos_all
```

File: hnn_core/cell.py
Comment (cjayb) [inline hnn_core/cell.py:323]: Along the lines of my other comments, could this be `'apical_section'`?

`The name of a section defining the apical orientation, onto which dipole moments are projected.`

**Review by cjayb — APPROVED** [review]

File: hnn_core/cell.py
Comment (jasmainak) [reply hnn_core/cell.py:323]: isn't "apical" specific to the default neuron? I like the use of term "projection" though. I'll try to incorporate it

File: hnn_core/cell.py
Comment (jasmainak) [reply hnn_core/cell.py:17]: Not really. This just projects along the apical dendrite which may or may not be oriented along the y direction (Neuron coordinates ... z direction in physical coordinates). 

There are the following two alternatives which I considered and that does assume the xy-plane:

```py
y_long = (h.y3d(1, sec=sect) - h.y3d(0, sec=sect)) * pos_all
```

and 

```py
y_long = (p_secs[sec_name]['sec_pts'][1][1] - p_secs['apical_trunk']['sec_pts'][0][1]) * pos_all
```

and get rid of `yscale` at all. That didn't work though because of numerical issues I suspect. The code uses a mix of Neuron and Python. At some point we should figure out what kind of rounding/precision Neuron actually uses. We could deliberately typecast in Python then to make things match and then remove it when we're confident it's correct

File: hnn_core/cell.py
Comment (jasmainak) [inline hnn_core/cell.py:345]: for the love of God, I can't figure out what this is doing. This is equivalent to doing:

section length * segment length * cos(theta)

But units of dipole is nAm, that is to say it is "length x current" and here we have length ** 2. Or am I missing something?

Comment (jasmainak) [inline hnn_core/cell.py:345]: this scaling is super mysterious to me. We're doing:

diff(section length * position vector of segment)
= section length * segment lengths

File: hnn_core/cell.py
Comment (cjayb) [reply hnn_core/cell.py:17]: > Not really. This just projects along the apical dendrite which may or may not be oriented along the y direction

But isn't index `1` in `p_secs[dipole_sec]['sec_pts'][1]` explicitly the y-coordinate? And `0` on the next line x. That's what I mean by "hard assumptions"---we're doing 2D projection, implicitly ignoring the z-orientation, which is assumed to be degenerate (flat).

File: hnn_core/cell.py
Comment (cjayb) [reply hnn_core/cell.py:323]: In my mind, "apical" refers to the single large dendrite sprouting from the top of the pyramid-formed soma (from the "apex" of the pyramid). Some L2/3 pyramidal cells have a bifurcation very close to the soma, but even then the two apical dendrites run in parallel, towards L1. The "textbook answer" is that the apical dendrite(s) is a defining characteristic of _all_ neocortical (indeed, also hippocampal) pyramidal cells. The terms 'proximal' and 'distal' are much more ambiguous in comparison.

File: hnn_core/cell.py
Comment (cjayb) [reply hnn_core/cell.py:345]: Yeah I was trying to make sense of that too... I think it's very ugly shorthand. The L and theta are constant, so the diff is effectively on the position, which is in units of section length (`[0, 1]`). This may have something to do with calculating the _dipole moment_ of current, but it all depends on what `Dipole.mod` and `Dipole_pp.mod` are actually doing... I'm digging into the details of HOC (trying to figure out my LFP-woes). I have plans to visit the "Dipole"-calculations, too, hoping this part of our code will make more sense!

Comment (jasmainak) [reply hnn_core/cell.py:345]: oh I get it now. `sect.L * np.diff(pos_all)` is the fractional length (in units of um) of that segment. Multiply that by the current in the segment and you get the dipole moment of that segment

File: hnn_core/cell.py
Comment (jasmainak) [reply hnn_core/cell.py:323]: humm ... so are you saying that you'd always want to compute the dipole along an apical dendrite? I'll change the name in that case

Comment (cjayb) [reply hnn_core/cell.py]: Yes. Even if we one day included a more morphologically detailed model, the dipole moment would still be calculated along the primary symmetry axis defined by the apical dendrite. You know how in MNE we have the "loose orientation constraint"? The "looseness" is fudge factor to cover for coreg and surface recon errors. The underlying _strong_ assumption is that there's one dominant dipole orientation per source space point. And this is the normal to the cortical (WM) surface _because of_ the apical dendrites.

File: hnn_core/cell.py
Comment (cjayb) [reply hnn_core/cell.py]: Indeed, but that's what I mean by having to understand the MOD-files better. Because then they must not be calculating the dipole moment, right? But rather the primary current density...

File: hnn_core/cell.py
Comment (cjayb) [reply hnn_core/cell.py]: Ah, my bad. All clear now!

File: hnn_core/cell.py
Comment (cjayb) [inline hnn_core/cell.py:19]:
```suggestion
    cos_theta = dict()
```

File: hnn_core/cell.py
Comment (cjayb) [inline hnn_core/cell.py:22]:
```suggestion
        cos_theta[sec_name] = np.dot(a, b) / (norm(a) * norm(b))
```

File: hnn_core/cell.py
Comment (cjayb) [inline hnn_core/cell.py:23]:
```suggestion
    return cos_theta
```

File: hnn_core/cell.py
Comment (cjayb) [inline hnn_core/cell.py:323]:
```suggestion
            The name of the section along which dipole moment is calculated.
```

WDYT?

**Review by cjayb — APPROVED** [review]
Final nitpicks, good to merge!

**Comment by jasmainak** [comment]
Addressed! Thanks for the careful reviews :)

File: hnn_core/cell.py
Comment (rythorpe) [reply hnn_core/cell.py]: I think you're right. I'm guessing `Qsum` is the the sum of all radial current entering the dendrite at a given point in a nrn section?

**Review by rythorpe — APPROVED** [review]

**Comment by cjayb** [comment]
Thanks for the much more readable code @jasmainak, merged after green light from @rythorpe

**Pull request merged**
